core/src/modules/ops/record.ts

```diff
@@ -293,159 +293,159 @@ export namespace RecordOps {
+          if (matchIdx >= 0) {
+            bulkCreates.splice(matchIdx, 1, bulkCreate);
+          } else {
+            bulkCreates.push(bulkCreate);
           }
 
-          // add new GrouparooRecord Properties to batch
-          let position = 0;
-          buildQueries: for (const value of h[key]) {
-            if (position > 0 && !property.isArray) {
-              throw new Error(
-                "cannot set multiple record properties for a non-array property"
-              );
-            }
+          position++;
+        }
 
-            const existingRecordProperty = existingRecordProperties.find(
-              (p) =>
-                p.recordId === record.id &&
-                p.propertyId === property.id &&
-                p.position === position
-            );
-            const { rawValue, invalidValue, invalidReason } =
-              await RecordPropertyOps.buildRawValue(
-                value,
-                property.type,
-                existingRecordProperty
-              );
-
-            bulkCreates.push({
-              id: existingRecordProperty
-                ? existingRecordProperty.id
-                : undefined,
-              recordId: record.id,
-              propertyId: property.id,
-              position,
-              rawValue,
-              invalidValue,
-              invalidReason,
-              state: "ready",
-              stateChangedAt: now,
-              confirmedAt: now,
-              startedAt: null,
-              valueChangedAt:
-                !existingRecordProperty ||
-                !existingRecordProperty.valueChangedAt ||
-                !existingRecordProperty.rawValue ||
-                rawValue !== existingRecordProperty.rawValue
-                  ? now
-                  : existingRecordProperty.valueChangedAt,
-              unique: property.unique,
-            });
-
-            position++;
-          }
+        // delete old properties we didn't update
+        existingRecordProperties
+          .filter(
+            (p) =>
+              p.recordId === record.id &&
+              p.propertyId === property.id &&
+              p.position >= position
+          )
+          .map((p) => bulkDeletes.where.id.push(p.id));
+      }
 
-          // delete old properties we didn't update
-          existingRecordProperties
-            .filter(
-              (p) =>
-                p.recordId === record.id &&
-                p.propertyId === property.id &&
-                p.position >= position
-            )
-            .map((p) => bulkDeletes.where.id.push(p.id));
-        }
+      recordOffset++;
+    }
 
-        recordOffset++;
+    // TODO: This select may be slower than allowing the upsert to throw like we used to
+    // However, the previous upsert method was really hard to deal with because it would throw and ruin/end the transaction
+    const duplicateRecordRawValues = bulkCreates
```

Comment:
As the comment says, we can't rely on `updateOnDuplicate` throwing like we used to as we need this method to succeed in-line